contributors/design-proposals/admission_control_extension.md

```diff
@@ -0,0 +1,594 @@
+6. Quota must still run after all validators are invoked. We may need to make quota extensible in the future.
+
+
+#### Initializers
+
+An initializer must allow some API clients to perceive creations prior to other apps. For backwards compatibility,
+uninitialized objects must be invisible to legacy clients. In addition, initializers must be recognized as
+participating in the initialization process and therefore a list of initializers must be populated onto each
+new object. Like finalizers, each initializer should perform an update and remove itself from the object.
+
+Every API object will have the following field added:
+
+```
+type ObjectMeta struct {
+    ...
+    // Initializers is a list of initializers that must run prior to this object being visible to
+    // normal clients. Only highly privileged users may modify this field. If this field is set,
+    // then normal clients will receive a 202 Accepted and a Status object if directly retrieved
+    // by name, and it will not be visible via listing or watching.
+    Initializers *Initializers `json:"initializers"`
+    ...
+}
+
+// Initializers tracks the progress of initialization.
+type Initializers struct {
+    // Pending is a list of initializers that must execute in order before this object is visible.
+    // When the last pending initializer is removed, and no failing result is set, the initializers
+    // struct will be set to nil and the object is considered as initialized and visible to all
+    // clients.
+    Pending []Initializer `json:"pending"`
+    // If result is set with the Failure field, the object will be persisted to etcd and then deleted,
+    // ensuring that other clients can observe the deletion.
+    Result *metav1.Status `json:"status"`
+}
+
+// Initializer records a single pending initializer. It is a struct for future extension.
+type Initializer struct {
+    // Name is the name of the process that owns this initializer
+    Name string `json:"name"`
+}
+```
+
+On creation, an admission controller defaults the initializers field (if nil) to a value from the system
+configuration that may vary by resource type. If initializers is set the admission controller will check
+whether the user has the ability to run the `initialize` verb on the current resource type, and reject
+the entry with 403 if not. This allows a privileged user to bypass initialization by setting `initializers`
+to the empty struct.
+
+Once created, an object is not visible to clients unless the following conditions are satisfied:
+
+1. The initializers field is null.
+2. The client provides a special option to GET, LIST, or WATCH indicating that the user wishes to see uninitialized objects.
+
+The apiserver that accepts the incoming creation should hold the response until the object is
+initialized or the timeout is exceeded. This increases latency, but allows clients to avoid breaking
```

Comment:
I guess it depends on whether we allow initializers to create resources referencing the uninitialized object (by allow, I mean, explicitly want to enable that).  I'm not sure that normal guarantees about orphan can or should apply to uninitialized resources.  The garbage collector definitely has to be aware of uninitialized resources (so that it doesn't GC one), but I'm not sure that synchronous clean up is required.